ocaml/typing/uniqueness_analysis.ml

```diff
@@ -437,19 +553,329 @@ end = struct
+  val assert_empty : t -> unit
+
+  val print : Format.formatter -> t -> unit
+end = struct
+  type t = Tag.Set.t
+
+  (* If we were to define a zero for this semiring, it would be
+     the set of all possible tags. *)
+
+  let empty = Tag.Set.empty
+
+  let choose t0 t1 = Tag.Set.inter t0 t1
+
+  let par t0 t1 = Tag.Set.union t0 t1
+
+  let seq t0 t1 = Tag.Set.union t0 t1
+
+  let learn_tag tag = Tag.Set.singleton tag
+
+  let extract_tags t = t
+
+  let assert_empty t = assert (Tag.Set.is_empty t)
+
+  let print ppf t =
+    let open Format in
+    fprintf ppf "%a" (Format.pp_print_list Tag.print) (Tag.Set.elements t)
+end
+
+module Overwrites : sig
+  (** This module collects the tags of overwrites. It is always sound
+      to assume we have overwritten with additional tags.
+
+      However, it is in general unsound if the user mutates the tag
+      after we have learned its nature. Then all bets are off and we
+      always fail if there is an overwrite that depends on this
+      information. *)
+
+  type t
+
+  type old_tag =
+    | Old_tag_unknown
+    | Old_tag_was of Tag.t
+    | Old_tag_mutated of access_order
+
+  type error =
+    | Changed_tag of
+        { old_tag : old_tag;
+          new_tag : Tag.t
+        }
+
+  exception Error of error
+
+  (** Unknown tag *)
+  val empty : t
+
+  (** Overwrite using a certain tag *)
+  val overwrite_tag : Tag.t -> t
+
+  (** Indicate a mutation (which invalidates all tags we have learned
+      and will learn). If there is also an overwrite, we fail. *)
+  val mutate_tag : t
+
+  (** Sequential composition *)
+  val seq : t -> t -> t
+
+  (** Non-deterministic choice *)
+  val choose : t -> t -> t
+
+  (** Parallel composition *)
+  val par : t -> t -> t
+
+  (** If we find out that a tag was mutated,
+      we need to promote this information to its children.
+      This is because a write to a mutable field can change
+      any tag that is reachable from the mutable field. *)
+  val promote_mutation_to_children : t -> t
+
+  (** We may not overwrite a tag we don't know.
+      At the end of the analysis, this function should
+      be called to ensure all overwrites are on known tags. *)
+  val check_no_remaining_overwritten_as : t -> unit
+
+  (** Accept the overwrites using tags that we have learned
+      from a pattern-match. *)
+  val match_with_learned_tags : Tag.Set.t -> t -> t
+
+  (** Assert that no overwrites were collected and fail else. *)
+  val assert_empty : t -> unit
+
+  val print : Format.formatter -> t -> unit
+end = struct
+  type tags =
+    { overwritten : Tag.Set.t Tag.Map.t;
+          (** The keys of the map are the tags of the overwrite.
+          When we learn new tags, we can remove keys from this map,
+          that correspond to the new tags. We then keep the learned
+          tags on the keys that did not match them to produce good
+          error messages. It is always sound to add elements to this map.
+          *)
+      was_mutated : bool
```

Comment:
What's this for? It looks to me that we would have `Tag_was_mutated` instead of using this bool...